### PR TITLE
Changes to the Noise class to store only local detectors.

### PR DESCRIPTION
**Pull request by tskisner**
* Remove the internal noise model from the focalplane class. Focalplanes have information about all detectors (not only the locally stored ones).

* Add methods to the Noise base class and the AnalyticNoise derived class to handle gather / scatter / redistribution of noise model data.

* Update the hdf5 I/O methods to use these new functions for when reading / writing.

* When redistributing observations, any metadata objects that define a redistribute() method will have that called.

* Update ElevationNoise, NoiseEstim, and FitNoiseModel operators to match the API changes to the noise model.

* Fix warnings throughout the code due to the use of "np.int", "np.int_", and "np.float" types.

### CODE REVIEW

**Review by keskitalo — APPROVED** [review]
This is a huge improvement to TOAST capabilities. It cuts the memory overhead for the noise model significantly and makes load balancing much more flexible.

File: src/toast/noise_sim.py
Comment (tskisner) [reply src/toast/noise_sim.py:183]: This function modifies the class members in place (returns None) and this return statement is a short circuit for the case of one process.

File: src/toast/noise_sim.py
Comment (tskisner) [reply src/toast/noise_sim.py:167]: This function returns None, and modifies the `out` dictionary in place.  The error is actually on line 150 where I returned this rather than just having an empty return.

File: src/toast/ops/noise_estimation.py
Comment (tskisner) [reply src/toast/ops/noise_estimation.py]: Fixed

File: src/toast/ops/elevation_noise.py
Comment (tskisner) [reply src/toast/ops/elevation_noise.py]: Sorry, habit- changed back, since it was an irrelevant style change.

File: src/toast/observation.py
Comment (tskisner) [reply src/toast/observation.py]: Done

**Comment by tskisner** [comment]
Just a note on this PR.  The CMBenv docker containers used to test extra dependencies on linux include an MPI enabled HDF5 / h5py which seems to have a problem.  I have run the unit tests in this branch locally in a conda environment with 4 processes using hdf5-mpi installed from conda packages.  Our cmb stack is in the process of being updated, but that should not hold up this PR.  Will merge once MacOS tests pass.

**Pull request merged**
